src/librustc_save_analysis/json_dumper.rs

```diff
@@ -10,73 +10,402 @@
 
 use std::io::Write;
 
+use rustc::hir::def_id::DefId;
 use rustc_serialize::json::as_json;
 
 use super::external_data::*;
 use super::dump::Dump;
 
 pub struct JsonDumper<'b, W: Write + 'b> {
     output: &'b mut W,
-    first: bool,
+    result: Analysis,
 }
 
 impl<'b, W: Write> JsonDumper<'b, W> {
     pub fn new(writer: &'b mut W) -> JsonDumper<'b, W> {
-        if let Err(_) = write!(writer, "[") {
-            error!("Error writing output");
-        }
-        JsonDumper { output: writer, first: true }
+        JsonDumper { output: writer, result: Analysis::new() }
     }
 }
 
 impl<'b, W: Write> Drop for JsonDumper<'b, W> {
     fn drop(&mut self) {
-        if let Err(_) = write!(self.output, "]") {
+        if let Err(_) = write!(self.output, "{}", as_json(&self.result)) {
             error!("Error writing output");
         }
     }
 }
 
 macro_rules! impl_fn {
-    ($fn_name: ident, $data_type: ident) => {
+    ($fn_name: ident, $data_type: ident, $bin: ident) => {
```

Comment:
What does the name `$bin` mean here? Is this `$bin` in the sense of "container" (i.e. the container that is receiving the values being converted from the input `data`)?

(Usually when a see the word "bin", my first thought is that it is short for "executable binary", like `/usr/bin`, so that's why I ask.)